contracts/v3-migration/OriginationControllerMigrate.sol

```diff
@@ -0,0 +1,472 @@
+    ) internal {
+        IERC20 asset = assets[0];
+
+        _repayLoan(asset, opData.oldLoanId, amounts[0]);
+
+        _initializeMigrationLoan(
+            opData.newLoanTerms,
+            opData.itemPredicates,
+            borrower,
+            opData.lender
+        );
+
+        // pull funds from the lender to repay the flash loan
+        IERC20(opData.newLoanTerms.payableCurrency).safeTransferFrom(opData.lender, address(this), opData.migrationAmounts.amountFromLender - opData.migrationAmounts.leftoverPrincipal);
+
+        if (opData.migrationAmounts.amountToBorrower > 0) {
+            // If new principal is greater than old loan repayment amount, send the difference to the borrower
+            IERC20(opData.newLoanTerms.payableCurrency).safeTransfer(borrower, opData.migrationAmounts.amountToBorrower);
+        }
+
+        // Make flash loan repayment
+        // Balancer requires a transfer back the vault
+        asset.safeTransfer(VAULT, amounts[0] + premiums[0]);
+    }
+
+    /**
+     * @notice Helper function to repay the V3 loan.
+     *
+     * @param payableCurrency              Payable currency for the loan terms.
+     * @param borrowerNoteId               ID of the borrowerNote for the loan to be repaid.
+     * @param repayAmount                  The amount to be repaid on V3
+     */
+    function _repayLoan(
+        IERC20 payableCurrency,
+        uint256 borrowerNoteId,
+        uint256 repayAmount
+    ) internal {
+        // approve LoanCoreV3 to take the total settled amount
+        payableCurrency.safeApprove(loanCoreV3, repayAmount);
+
+        // repay V3 loan, this contract receives the collateral
+        IRepaymentControllerV3(repaymentControllerV3).repay(borrowerNoteId);
+    }
+
+    /**
+     * @notice Helper function to initialize the new v4 loan.
+     *
+     * @param newTerms                  The terms of the v4 loan.
+     * @param itemPredicates            The predicates for the v4 loan.
+     * @param borrower_                 The address of the borrower.
+     * @param lender                    The address of the lender.
+     *
+     * @return newLoanId                The ID of the new loan.
+     */
+    function _initializeMigrationLoan(
+        LoanLibrary.LoanTerms memory newTerms,
+        LoanLibrary.Predicate[] memory itemPredicates,
+        address borrower_,
+        address lender
+    ) internal returns (uint256 newLoanId) {
+        // transfer collateral to LoanCore
+        IERC721(newTerms.collateralAddress).transferFrom(address(this), address(loanCore), newTerms.collateralId);
+
+        // all v4 loan data fees are set to zero for migrations
+        LoanLibrary.FeeSnapshot memory feeSnapshot = LoanLibrary.FeeSnapshot({
+            lenderDefaultFee: 0,
+            lenderInterestFee: 0,
+            lenderPrincipalFee: 0
+        });
+
+        // create loan in LoanCore
+        newLoanId = loanCore.startLoan(lender, borrower_, newTerms, 0, 0, feeSnapshot);
```

Comment:
I didnt want to have fees be an impact on a user who is deciding whether to rollover (migrate) from v3-v4 or just close their loan on v3. It didnt seem right and is a big gas cost. I didnt want anything to disincentivize someone from migrating to v4. It didnt seem right to add it because we changed the protocol. If fees are turned on for V4 origination, I didnt want that connected to v3 which is a zero fee enviroment as it currently stands, it would not incentives rollovers and could result in loans being closed.

I think we have the contract space now (~0.25 KB) that I removed passing around the predicates data unnecessarily, so if you feel strongly it should be in there I can add it back in. I just wanted to provide some reasoning behind that decision.